Partial Essential Boundary Conditions in NonlinearForm

rcarson3 opened the pull request:
This adds support for per boundary to set what component(s) of the DOFs as being essential DOFs. This is particularly useful for certain types of problems such as those in commonly found in solid mechanics where it's useful to have varying # of the DOF components constrained per surface/boundary.

I've taken the code here from what we've been using within the exaconstit-dev branch for the past few years. A limitation of the current implementation presented in this PR is that it has not been generalized to nonconforming meshes. Although, it might be trivial to extend it to that case given the code for this use case is largely shared with the code already in MFEM for getting essential vdofs. However, I don't really understand that use case well enough to know how the nonconforming case differs.
<!--GHEX{"author":"rcarson3","assignment":"2022-02-25T02:59:11","editor":"tzanio","id":2849,"reviewers":["mlstowell","vladotomov","jandrej"],"merge":"2022-04-12T01:59:11","approval":"2022-04-05T01:59:11"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2849](https://github.com/mfem/mfem/pull/2849) | @rcarson3 | @tzanio | @mlstowell + @vladotomov + @jandrej | 2/24/22 | ⌛due 4/4/22 | ⌛due 4/11/22 | |
<!--ELBATXEHG-->

rcarson3 committed:
Changes from ExaConstit mfem branch that are useful for having varying partial Essential DOFs per boundary attribute

rcarson3 committed:
make style

rcarson3 committed:
fix doxygen comments

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

jandrej commented on fem/fespace.cpp:
```diff
@@ -561,6 +561,68 @@ void FiniteElementSpace::GetEssentialVDofs(const Array<int> &bdr_attr_is_ess,
+         {
+            cmpID_rows = cmpID_rows && ( component(id, j) > -1);
+         }
+         if (cmpID_rows) // same as srw component id system
```
what is srw?

rcarson3 replied:
Sorry I missed that comment. I thought I'd gotten most of them. It's the initials of one the initial contributors to ExaConstit.

jandrej commented on fem/nonlinearform.cpp:
```diff
@@ -53,6 +53,22 @@ void NonlinearForm::SetEssentialBC(const Array<int> &bdr_attr_is_ess,
    }
 }
 
+void NonlinearForm::SetEssentialBCPartial(const Array<int> &bdr_attr_is_ess,
```
What does `Partial` mean in this case? The term is ambiguous. I would opt for keeping the same name `NonlinearForm::SetEssentialBC`. The added argument `bdr_component` and class method comment make it clear that this will work different from the non component version.

rcarson3 replied:
`Partial` was just the name we'd had picked back in the day. It was supposed to just signify that you're not constrained to either constraining all dofs or only 1 dof. Instead, you could constrain multiple DOFs at a time for a boundary. I really don't have strong preferences to keeping the name though.

jandrej commented on fem/nonlinearform.hpp:
```diff
@@ -146,6 +146,19 @@ class NonlinearForm : public Operator
        have zero entries at the essential true dofs. */
    void SetEssentialBC(const Array<int> &bdr_attr_is_ess, Vector *rhs = NULL);
 
+   /// Specify essential boundary conditions. If vdim > 1 then one can specify
+   /// per boundary attribute which components on a boundary are essential by
+   /// assigning the component ID to its location in bdr_component array
+   /// The bdr_component has dimensions number boundary attributes x vdim
+   /** This method calls FiniteElementSpace::GetEssentialTrueDofs() and stores
+       the result internally for use by other methods. If the @a rhs pointer is
+       not NULL, its essential true dofs will be set to zero. This makes it
+       "compatible" with the output vectors from the Mult() method which also
+       have zero entries at the essential true dofs. */
+   void SetEssentialBCPartial(const Array<int> &bdr_attr_is_ess,
+                              const Array2D<int> &bdr_component,
+                              Vector *rhs);
```
OK for me only because the existing class already implements this the same way. For naming, see above.

jandrej reviewed (commented):
I'm wondering what these additions bring to the table. Can't I just loop over boundary attributes separately with the existing `FiniteElementSpace::GetEssentialTrueDofs` and 3 more lines of code instead of adding a whole new interface?

For me this is a pass.

tzanio commented:
Mention in `CHANGELOG`?

rcarson3 commented:
> I'm wondering what these additions bring to the table. Can't I just loop over boundary attributes separately with the existing `FiniteElementSpace::GetEssentialTrueDofs` and 3 more lines of code instead of adding a whole new interface?
> 
> For me this is a pass.

@jandrej so using your way of doing it, you would need to call the existing `FiniteElementSpace::GetEssentialTrueDofs`  vdim x number of boundary attributes and combine the resulting matrices multiple times to achieve the same results. Alternatively, you could reduce the number of calls by calling it only vdim times but that route would require you to have some book-keeping about which boundary attribute is constrained to a given component. I would say that this method does simplify things for the user as it's much easier to keep track of what boundaries are constrained and for what DOFs.

It might be that we could make use of the existing code to do this rather than duplicating things, where we could do the book-keeping internally for the user when they call the `FiniteElementSpace` methods added in this PR. However, it might require a few small changes to `FiniteElementSpace::GetEssentialVDofs` aka we might want to add a flag which allows the inputted `ess_vdofs` to not be overwritten.

pazner commented on fem/pfespace.hpp:
```diff
@@ -355,12 +355,36 @@ class ParFiniteElementSpace : public FiniteElementSpace
                                   Array<int> &ess_dofs,
                                   int component = -1) const;
 
+   /** @brief Mark degrees of freedom associated with boundary elements with
+       the specified boundary attributes (marked in 'bdr_attr_is_ess').
+       For spaces with 'vdim' > 1, the 'component' parameter can be used
+       to restrict the marked tDOFs per boundary to the specified components.
+       If vdim > 1 then one can specify per boundary attribute which components
+       on a boundary are essential by assigning the component ID to its location
+       in component array.
+       The component has dimensions number of boundary attributes x vdim. */
+   virtual void GetEssentialVDofs(const Array<int> &bdr_attr_is_ess,
+                                  Array<int> &ess_vdofs,
+                                  const Array2D<int> &component) const;
+
    /** Get a list of essential true dofs, ess_tdof_list, corresponding to the
        boundary attributes marked in the array bdr_attr_is_ess. */
    virtual void GetEssentialTrueDofs(const Array<int> &bdr_attr_is_ess,
                                      Array<int> &ess_tdof_list,
                                      int component = -1);
 
+   /** @brief Get a list of essential true dofs, ess_tdof_list, corresponding to the
+       boundary attributes marked in the array bdr_attr_is_ess.
+       For spaces with 'vdim' > 1, the 'component' array can be used
+       to restricts the marked tDOFs per boundary to the specified components.
+       If vdim > 1 then one can specify per boundary attribute which components
+       on a boundary are essential by assigning the component ID to its location
+       in component array.
+       The component has dimensions number of boundary attributes x vdim. */
```
Ditto about the "brief" docs.

pazner reviewed (commented):
Some minor Doxygen formatting comments…

vladotomov commented:
I'm not convinced that a new interface is needed. I achieve this in Laghos in very few lines:
```
Array<int> ess_tdofs, ess_bdr(pmesh->bdr_attributes.Max()), dofs_list;
for (int d = 0; d < pmesh->Dimension(); d++)
{
   // Attributes 1/2/3 correspond to fixed-x/y/z boundaries,
   // i.e., we must enforce v_x/y/z = 0 for the velocity components.
   ess_bdr = 0; ess_bdr[d] = 1;
   H1FESpace.GetEssentialTrueDofs(ess_bdr, dofs_list, d);
   ess_tdofs.Append(dofs_list);
}
```
Sure the proposed would be slightly more flexible, but useful for very specific cases, and can be achieved with minimal bookkeeping from the user with the existing interface.

rcarson3 commented on fem/fespace.cpp:
```diff
@@ -561,6 +561,68 @@ void FiniteElementSpace::GetEssentialVDofs(const Array<int> &bdr_attr_is_ess,
+               {
+                  if (component(id, k) != -1)   // -1 means inactive component
+                  {
+                     // component(id, k) should be equal to k here
```
We'd just been trying to keep things consistent with how the original function works as it took an `int` argument earlier. A bool would probably work just as well.

rcarson3 commented:
> I'm not convinced that a new interface is needed. I achieve this in Laghos in very few lines: 
```c++
Array<int> ess_tdofs, ess_bdr(pmesh->bdr_attributes.Max()), dofs_list;
for (int d = 0; d < pmesh->Dimension(); d++)
{
   // Attributes 1/2/3 correspond to fixed-x/y/z boundaries,
   // i.e., we must enforce v_x/y/z = 0 for the velocity components.
   ess_bdr = 0; ess_bdr[d] = 1;
   H1FESpace.GetEssentialTrueDofs(ess_bdr, dofs_list, d);
   ess_tdofs.Append(dofs_list);
}
```
> Sure the proposed would be slightly more flexible, but useful for very specific cases, and can be achieved with minimal bookkeeping from the user with the existing interface.

@vladotomov it would probably be worth having a webex discussion about this with the other reviewers as I would say the book keeping required for problems with non-trivial BCs or those with evolving BCs using the existing interface is not user friendly. I can set that up a webex discussing if @jandrej , @v-dobrev or @pazner are interested in one as well.

tzanio commented:
Update on the current status: after internal discussion we decided to add the new API functions, but minimize the code duplication in the implementation. @rcarson3 will update the PR accordingly.

rcarson3 committed:
Removed duplicated code to address reviewers comments
Removed the more or less duplicated FESpace::GetEssentialVDofs and moved the necessary book-keeping portions of things to GetEssentialTrueDofs
Additionally, I added an overwrite optional variable (defaulted to true) to GetEssentialVDofs in-order to allow one to overwrite the values of ess_vdofs if they'd like.
I still need to add some unit tests for the functionality though.

rcarson3 committed:
Merge remote-tracking branch 'origin/master' into feature/partial_ess_bcs

rcarson3 committed:
make style

rcarson3 committed:
fix build issue noted on github ci

rcarson3 committed:
add unit test for fespace get essential tdofs

rcarson3 commented:
@vladotomov ,  @jandrej , @mlstowell , and @pazner  I believe this should be good to review again. I believe I've addressed the earlier reviewer comments by eliminating the code duplication related to `FiniteElementSpace::GetEssentialVDofs`. The new `FiniteElementSpace::GetEssentialTrueDofs` variation now does booking keeping on its own end of things and makes use of the previous version of `FiniteElementSpace::GetEssentialVDofs`. I will mention that for `FiniteElementSpace::GetEssentialVDofs` I did add an additional function argument that tells the function whether or not it should overwrite the incoming `ess_vdofs` argument. This new argument has a default value such that the old behavior of the function is retained. So, it should not breaking any existing code.

Also the doxygen comments should be updated to address earlier comments regarding them.

rcarson3 commented on fem/fespace.hpp:
```diff
@@ -775,11 +775,16 @@ class FiniteElementSpace
        For spaces with 'vdim' > 1, the 'component' parameter can be used
-       to restricts the marked vDOFs to the specified component. */
+       to restricts the marked vDOFs to the specified component.
+       If overwrite is set to false then values in ess_vdofs are preserved
+       and not reset. However, the assumption here is that ess_vdofs is set to
+       the correct size already.*/
    virtual void GetEssentialVDofs(const Array<int> &bdr_attr_is_ess,
                                   Array<int> &ess_vdofs,
-                                  int component = -1) const;
+                                  int component = -1,
+                                  bool overwrite = true) const;
```
Yes, it's just away to avoid having `ess_vdofs` being recreated every time this function is called and the need for another `Array` to keep track of the accumulated values. If you think improved documentation would help here to make that more clear I can do that.

rcarson3 commented on fem/nonlinearform.hpp:
```diff
@@ -146,6 +146,24 @@ class NonlinearForm : public Operator
+       in the bdr_component array.
+       The bdr_component has dimensions number of boundary attributes x vdim
+
+       This method calls FiniteElementSpace::GetEssentialTrueDofs() and stores
+       the result internally for use by other methods. If the @a rhs pointer is
+       not NULL, its essential true dofs will be set to zero. This makes it
+       "compatible" with the output vectors from the Mult() method which also
+       have zero entries at the essential true dofs. */
+   void SetEssentialBC(const Array<int> &bdr_attr_is_ess,
```
I could mention that here and in the other `SetEssentialBC` function right above.

rcarson3 commented on fem/pfespace.cpp:
```diff
@@ -1059,6 +1061,58 @@ void ParFiniteElementSpace::GetEssentialTrueDofs(const Array<int>
+                                                 Array<int> &ess_tdof_list,
+                                                 const Array2D<bool> &component)
+{
+   MFEM_ASSERT(component.NumCols() == vdim,
+               "Number of columns of component was not equal to ParFESpace vdim");
+   MFEM_ASSERT(component.NumRows() == bdr_attr_is_ess.Size(),
+               "Number of rows of component was not equal to bdr_attr_is_ess.Size()");
+
+   Array<int> ess_dofs, true_ess_dofs, bdr_attr_is_ess_single_comp;
+   bdr_attr_is_ess_single_comp.SetSize(bdr_attr_is_ess.Size());
+
+   for (int i = 0; i < vdim; i++)
+   {
+      // Only overwrite ess_vdofs on first iteration
+      // all other iterations we want to preserve values of
+      // ess_vdofs.
+      const bool overwrite = (i == 0) ? true : false;
+      bdr_attr_is_ess_single_comp = 0;
+      for (int j = 0; j < bdr_attr_is_ess.Size(); j++)
+      {
+         if (bdr_attr_is_ess[j] && component(j, i))
+         {
+            bdr_attr_is_ess_single_comp[j] = bdr_attr_is_ess[j];
+         }
+      }
+      GetEssentialVDofs(bdr_attr_is_ess_single_comp, ess_dofs, i, overwrite);
+   }
```
I'll try and see in what ways I can't refactor this to simplify things further.

tzanio commented:
ping: @rcarson3

vladotomov commented:
What is the state of this? Do you still want to merge it @rcarson3?

rcarson3 commented:
> What is the state of this? Do you still want to merge it @rcarson3?

@vladotomov sorry for a lack of movement on this. The project I was working on this for shifted gears shortly around my last set of comments. So, I just never got back to addressing the last set of comments. I'll see if I can't address all of Julian's comments next week when my schedule frees up just a little bit.

rcarson3 committed:
Merge branch 'master' into feature/partial_ess_bcs

rcarson3 committed:
add small guard against variable order H1 space in 2d component getessential true dofs for pfes

rcarson3 committed:
fix missing override

rcarson3 committed:
update copyright year

rcarson3 committed:
Minor doxygen update to address reviewer comments

rcarson3 committed:
Slight refactor of GetEssentialTrueDofs to reuse code in both FES and PFES classes
Claude helped generate the docs / do the refactor as I was lazy...
However, the unit tests pass and the amount of common code between the two cases has now been reduced which is always a good thing...

rcarson3 committed:
astyle fix based on CI git diff...

rcarson3 committed:
Merge branch 'master' into feature/partial_ess_bcs

rcarson3 committed:
Merge branch 'master' into feature/partial_ess_bcs

rcarson3 committed:
Merge branch 'master' into feature/partial_ess_bcs

rcarson3 committed:
Merge branch 'master' into feature/partial_ess_bcs

rcarson3 committed:
Merge branch 'master' into feature/partial_ess_bcs

rcarson3 committed:
Merge branch 'master' into feature/partial_ess_bcs

rcarson3 committed:
Merge branch 'master' into feature/partial_ess_bcs

jandrej reviewed (commented):
See https://github.com/mfem/mfem/pull/2849#pullrequestreview-893963762
